File: fsharp/e2e/TestHarness.fs
module CopilotSDK.Supercharged.FSharp.E2E.TestHarness

open System
open System.Diagnostics
open System.Runtime.InteropServices
open System.Text.RegularExpressions

/// Manages the replaying CAPI proxy process for E2E tests.
/// Spawns `npx tsx ../../test/harness/server.ts` and parses the listening URL.
type ReplayProxy() =
    let mutable proc: Process option = None
    let mutable proxyUrl: string option = None

    /// Start the replay proxy and return its URL (e.g. "http://localhost:PORT").
    member _.StartProxy() : string =
        match proxyUrl with
        | Some url -> url
        | None ->

        let serverPath =
            IO.Path.GetFullPath(
                IO.Path.Combine(__SOURCE_DIRECTORY__, "..", "..", "test", "harness", "server.ts"))

        let workingDir = IO.Path.GetDirectoryName(serverPath)
        let isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows)

        let psi = ProcessStartInfo()
        if isWindows then
            psi.FileName <- "cmd"
            psi.Arguments <- sprintf "/c npx tsx \"%s\"" serverPath
        else
            psi.FileName <- "npx"
            psi.Arguments <- sprintf "tsx \"%s\"" serverPath

        psi.WorkingDirectory <- workingDir
        psi.UseShellExecute <- false
        psi.RedirectStandardOutput <- true
        psi.RedirectStandardError <- false
        psi.CreateNoWindow <- true

        let p = Process.Start(psi)
        proc <- Some p

        // Read first line to get listening URL
        let line = p.StandardOutput.ReadLine()
        if String.IsNullOrEmpty(line) then
            p.Kill()
            failwith "Failed to read proxy URL from stdout"

        let m = Regex.Match(line, @"Listening: (http://[^\s]+)")
        if not m.Success then
            p.Kill()
            failwithf "Unexpected proxy output: %s" line

        let url = m.Groups.[1].Value
        proxyUrl <- Some url
        url

    /// Stop the replay proxy process.
    member _.StopProxy() =
        proc |> Option.iter (fun p ->
            try
                if not p.HasExited then p.Kill()
                p.WaitForExit()
                p.Dispose()
            with _ -> ())
        proc <- None
        proxyUrl <- None

    /// The proxy URL, or None if not started.
    member _.Url = proxyUrl

    interface IDisposable with
        member this.Dispose() = this.StopProxy()

File: ocaml/e2e/test_harness.ml
(* E2E test harness for the OCaml SDK.
   Spawns the shared replay proxy, parses the Listening URL,
   and provides helpers for E2E tests. *)

let server_path =
  let dir = Filename.dirname (Filename.dirname Sys.argv.(0)) in
  Filename.concat (Filename.concat (Filename.concat dir "..") "test")
    (Filename.concat "harness" "server.ts")

type proxy = {
  pid : int;
  url : string;
}

let start_proxy () =
  let abs_server =
    (* Navigate from ocaml/ (cwd) to repo root test/harness/server.ts *)
    let repo_root = Filename.dirname (Sys.getcwd ()) in
    Filename.concat repo_root
      (Filename.concat "test" (Filename.concat "harness" "server.ts"))
  in
  let cwd = Filename.dirname abs_server in
  let cmd = Printf.sprintf "cd %s && npx tsx %s" (Filename.quote cwd) (Filename.quote abs_server) in
  let ic = Unix.open_process_in cmd in
  let line = input_line ic in
  let url =
    let re = Str.regexp {|Listening: \(http://[^ ]+\)|} in
    if Str.string_match re line 0 then Str.matched_group 1 line
    else failwith (Printf.sprintf "Could not parse proxy URL from: %s" line)
  in
  (* We need to keep the channel alive so the process stays running *)
  let pid = Unix.getpid () in  (* placeholder -- proxy runs in background *)
  ignore pid;
  Printf.printf "[ocaml-e2e] Proxy listening at %s\n%!" url;
  Unix.putenv "COPILOT_API_URL" url;
  { pid = 0; url }

let stop_proxy proxy =
  ignore proxy;
  (* The proxy process gets cleaned up when the pipe closes *)
  ()

let proxy_url proxy = proxy.url
